From 4fea3d796af34ef085aa638ecc86ce5867d2c55d Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Fri, 27 Apr 2012 09:26:10 +0300
Subject: [PORT FROM MAIN] netd: Reset the InterfaceTable upon failure

BZ: 24583

Since the MOBILE_HIPRI is not a default connection,
ConnectivityService adds the route information of the
established HIPRI connection as a secondary one. Adding
of the route to the SecondaryTable failed. Eventhough the
adding of route information fails, interface name maintained in
the mInterfaceTable structure is not reset. When the call to
enableNat is made by Tethering service, NatController first
checks whether it's a secondary connection. Since the interface
table in the SecondaryTableController is not reset, NatController
tries to do the operations on the invalid secondary interface
resulting in exception thrown and eventually resulting in stopping
the tethering service.

This patch resets the mInterfaceTable upon failure.

Change-Id: I467238eb1da7d057ba7cc5f16c4a51bf6b22a04d
Orig-Change-Id: I7e38b00564cb86e826ae149ad34917881929162f
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Reviewed-on: http://android.intel.com:8080/82132
Reviewed-by: Fremaux, Pascal <pascal.fremaux@intel.com>
Reviewed-by: Ravindran, Arun <arun.ravindran@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Lebsir, SamiX <samix.lebsir@intel.com>
Tested-by: Lebsir, SamiX <samix.lebsir@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 SecondaryTableController.cpp |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/SecondaryTableController.cpp b/SecondaryTableController.cpp
index ce23d28..3e949aa 100644
--- a/SecondaryTableController.cpp
+++ b/SecondaryTableController.cpp
@@ -120,6 +120,7 @@ int SecondaryTableController::modifyRoute(SocketClient *cli, const char *action,
         ALOGE("ip route %s failed: %s route %s %s/%d via %s dev %s table %d", action,
                 IP_PATH, action, dest, prefix, gateway, iface, tableIndex+BASE_TABLE_NUMBER);
         errno = ENODEV;
+        mInterfaceTable[tableIndex][0] = 0;
         cli->sendMsg(ResponseCode::OperationFailed, "ip route modification failed", true);
         return -1;
     }
-- 
1.7.4.1

